### DISCUSSION TITLE
Suggested task for Outreachy Dec 22: Task 2 - Keras to MDF

### DISCUSSION THREAD

**Discussion by pgleeson** [ModECI/MDF#345] [Ideas]
A suggested set of tasks for applicants to the **Outreachy Winter 2022** call.

**Motivation:** Keras is a simple API for creating neural network models. We want people to be able to create a model using Keras, train it, and then export it to MDF to use it with any MDF compatible tools.

1) Install MDF locally.
2) Install [Keras](https://keras.io/)
3) Create a simple feedforward network with a few layers in Keras. this can be a "dummy" network, or a trained network based on an existing Keras tutorial.
4) Create another python script, which loads the **trained** network (i.e. with the final connection weights) and parses the structure and exports it to the equivalent network in MDF (i.e. with one node per layer which implements the activation function and edges for the connections which contain the trained weights)
5) Create a simple Jupyter notebook which demonstrates the conversion, illustrates that the keras version and the MDF version produce the same results, and shows a graphical output of the MDF equivalent.
6) Email us a link to your work.

**Comment by AniketARS** [comment]
Hi @pgleeson, in point 4 `edges for the connections which contain the trained weights` what does this means if we have to calculate `X@W+b` on the edge? because i've created parameters on edge and evaluating graph but edge is not evaluating!!

**Reply by pgleeson** [reply]
@AniketARS The edge can have an array for the weight matrix (e.g. https://github.com/ModECI/MDF/blob/main/examples/MDF/RNN/RNNs.yaml#L110), but it's probably best if the bias is in the target node and the calculation with that is done in the node.

**Reply by AniketARS** [reply]
@pgleeson edge weight seems not to work correctly because I tested with input shape (1, 2) and edge weight shape (2, 2) and getting result with shape (2, 2) which should be (1, 2) as it is `matmul` but it doing broadcast multiplication

**Reply by pgleeson** [reply]
@AniketARS This is the location where the weight on an edge is incorporated:
https://github.com/ModECI/MDF/blob/c4c334dac47989060526446d335ff08004c4a7dd/src/modeci_mdf/execution_engine.py#L1186
I guess we should replace that by a check on the type of the input/weight and potentially call numpy.matmul (though the parameters could be Tensors also...).

For now, you could incorporate the weight into the target node also, and do the multiplication explicitly there.

**Reply by AniketARS** [reply]
@pgleeson if I can work on about issue, for which you suggesting of adding check?

**Reply by pgleeson** [reply]
@AniketARS Last comment wasn't very clear..

**Reply by AniketARS** [reply]
Sorry for that @pgleeson I was asking to solve the issue you mentioned in https://github.com/ModECI/MDF/discussions/345#discussioncomment-3907636

**Reply by pgleeson** [reply]
Thanks @AniketARS, but the ultimate solution for that will depend on a number of other conversations we're having in the project now, specifically about how much freedom should be given to parameter types in MDF, i.e. whether the size/shape/type of input/output parameters should be specified in advance or be free for the user to choose. In the latter case it means that we will have to support scalars as well as arrays/tensors in expressions like `weight * value` and deal with (and document!) the various options accordingly.

So this would potentially be a task for the successful applicant to undertake in discussion with the rest of the developers during the internship. As suggested, for now you could bundle the weight into the target node, and ensure it's being handled correctly there.

**Comment by Monsurat-Onabajo** [comment]
Hello, good evening, i was working on the second task, i have installed the modeci_mdf library but i discovered i cant plot the graphs, i kept getting this error 
![IMG_20221018_221342](https://user-images.githubusercontent.com/102332122/196550245-fccd7d03-2e41-479b-94c4-4398de6c6d31.jpg)

**Reply by AniketARS** [reply]
hello @Monsurat-Onabajo , I also faced this problem a lot when I was working on notebook, this happens when we use some `notations` can't be represented by `graphviz ‘dot‘ engine,` because that library is used to plot model. Hence it is a more of a `graphviz` issue rather than being of `modeci_mdf`. You can use `.to_yaml()` or `.to_json()` method for visualization as it did to work around this.
Hope this will help you.

**Reply by Monsurat-Onabajo** [reply]
Ok, thank you very much, i will try that

**Reply by pgleeson** [reply]
See also comment here re graphviz: https://github.com/ModECI/MDF/discussions/224#discussioncomment-2535820

**Reply by Monsurat-Onabajo** [reply]
Thank you, i will try that

**Reply by Monsurat-Onabajo** [reply]
Hello, i have downloaded graphiz and everything is working perfectly fine but when i try using it on the model i created, i kept getting error but after doing research i discovered it is because of the > sign that is in two of my nodes, i have tried working around it but i kept getting error this error, is there anything i can do because my graph is working fine and giving me the expected result, it is the visualization that is giving me issues
![IMG_20221020_123631](https://user-images.githubusercontent.com/102332122/196939882-06791876-3ddf-46a5-a8b9-14fa64076f8e.jpg)

**Reply by pgleeson** [reply]
@Monsurat-Onabajo Glad you got it installed. 

Yes, the > might well mess up the generation of the graph via graphviz, since the format uses <> for embedded html. Where do you have the > in your model, is it a function expression? Technically those are possible to use (e.g. c.value = a>b), but may be restricted in future as it could be unclear what's meant when a and b are lists, arrays, tensors, etc. You could look at using a ParameterCondition also:
https://github.com/ModECI/MDF/blob/c4c334dac47989060526446d335ff08004c4a7dd/examples/MDF/RNN/generate_iaf.py#L65
which is supported in graph generation: https://github.com/ModECI/MDF/blob/main/examples/MDF/RNN/iaf.png

**Reply by Monsurat-Onabajo** [reply]
Thank you, i am using the `>` as a function expression but i will try your suggestions out

**Reply by Monsurat-Onabajo** [reply]
Thank you very much for your help, the parameterCondition() worked but even tho it is predicting same thing that the keras model predicted, the prediction probability of the mdf_model was off by ±0.04  but i was able to plot the graph, can i leave it like that? And can i upload the project on github?

**Reply by pgleeson** [reply]
@Monsurat-Onabajo Sounds good. Just commit it to a GH repo on your own account and mail me a link (or @ me and @sanjayankur31).

**Reply by Monsurat-Onabajo** [reply]
Hello, i have emailed you the link to my second task, thank you for all your help

**Reply by Monsurat-Onabajo** [reply]
Hello, i was going through my github today and i noticed that my codes for this project has been cloned by seven people, what can i do? And my code has not been checked yet so i cant make a submission

**Reply by pgleeson** [reply]
Don't worry, we'll look over your code and keep an eye out for any suggestion of plagiarism. Everyone working this task should start with a different Keras network anyway, and take a different approach to converting it.

**Reply by Monsurat-Onabajo** [reply]
Ok, thank you very much

**Reply by Monsurat-Onabajo** [reply]
Hello, As regards the completion of my final Outreachy application, as one of the requirement, I was asked to contact my mentor for a timeline of work to be accomplished on various projects and tasks to be concluded at each stage.

**Reply by pgleeson** [reply]
See https://github.com/ModECI/MDF/discussions/346#discussioncomment-3960400

**Comment by kusanele** [comment]
Good day, are we permitted to use our own dataset (upload data from our local machine to the repo) to train the model?

**Reply by pgleeson** [reply]
Sure, that's fine, as long as the code can be rerun. It shouldn't take too long to train though.
